File: apps/client/src/schemas/training-records.ts
import { z } from "zod"

import {
  practiceAnswerSchema,
  practiceDifficultySchema,
  practiceEvaluationSchema,
  practiceFollowUpReferenceAnswerStateSchema,
  practiceMainReferenceAnswerStateSchema,
  practiceQuestionTypeSchema,
  practiceQuestionSourceSchema,
  practiceRecommendationSchema,
} from "@/schemas/practice"

const uuidSchema = z.uuid()
const dateTimeSchema = z.iso.datetime({ offset: true })
const interactionLanguageSchema = z.enum(["zh-CN", "en"])

export const targetedPracticeTrainingRecordTargetRoleSchema = z
  .object({
    id: uuidSchema,
    title: z.string().min(1),
    company: z.string().nullable(),
  })
  .strict()

export const targetedPracticeTrainingRecordSetupSchema = z
  .object({
    source: practiceQuestionSourceSchema,
    prioritizeWeaknesses: z.boolean(),
  })
  .strict()

export const targetedPracticeTrainingRecordQuestionSchema = z
  .object({
    questionCardId: uuidSchema,
    prompt: z.string().min(1),
    questionType: practiceQuestionTypeSchema,
    difficulty: practiceDifficultySchema,
    assessedCapabilities: z.array(z.string()),
    isSaved: z.boolean(),
    isMarkedWeak: z.boolean(),
    referenceAnswer: practiceMainReferenceAnswerStateSchema,
  })
  .strict()

export const targetedPracticeTrainingRecordFollowUpSchema = z
  .object({
    questionId: uuidSchema,
    prompt: z.string().min(1),
    order: z.number().int().min(1).max(2),
    askedAt: dateTimeSchema,
    answer: practiceAnswerSchema.nullable(),
    referenceAnswer: practiceFollowUpReferenceAnswerStateSchema,
  })
  .strict()

export const targetedPracticeTrainingRecordEvaluationSchema = practiceEvaluationSchema

export const targetedPracticeTrainingRecordReviewSchema = z
  .object({
    overallPerformance: z.string().min(1),
    highlights: z.array(z.string()),
    mainIssues: z.array(z.string()),
    improvementSuggestions: z.array(z.string()),
    reusableAnswerStructure: z.array(z.string()),
    exposedWeaknesses: z.array(z.string()),
  })
  .strict()

export const targetedPracticeTrainingRecordAttemptSchema = z
  .object({
    attemptId: uuidSchema,
    attemptNumber: z.number().int().min(1),
    retryOfAttemptId: uuidSchema.nullable(),
    completedAt: dateTimeSchema.nullable(),
    question: targetedPracticeTrainingRecordQuestionSchema,
    mainAnswer: practiceAnswerSchema.nullable(),
    followUps: z.array(targetedPracticeTrainingRecordFollowUpSchema),
    evaluation: targetedPracticeTrainingRecordEvaluationSchema.nullable(),
    review: targetedPracticeTrainingRecordReviewSchema.nullable(),
    recommendation: practiceRecommendationSchema.nullable(),
  })
  .strict()

export const targetedPracticeTrainingRecordDetailResponseSchema = z
  .object({
    recordId: uuidSchema,
    kind: z.literal("targetedPractice"),
    status: z.enum(["completed", "endedEarly", "partiallyCompleted"]),
    language: interactionLanguageSchema,
    startedAt: dateTimeSchema,
    endedAt: dateTimeSchema,
    durationSeconds: z.number().int().nonnegative(),
    targetRole: targetedPracticeTrainingRecordTargetRoleSchema,
    setup: targetedPracticeTrainingRecordSetupSchema,
    attempts: z.array(targetedPracticeTrainingRecordAttemptSchema).min(1),
    exposedWeaknesses: z.array(z.string()),
    recommendation: practiceRecommendationSchema.nullable(),
  })
  .strict()
  .superRefine((record, context) => {
    if (record.endedAt < record.startedAt) {
      context.addIssue({
        code: "custom",
        message: "endedAt cannot be before startedAt",
        path: ["endedAt"],
      })
    }
    const attemptNumbers = record.attempts.map((attempt) => attempt.attemptNumber)
    if (!attemptNumbers.every((number, index) => number === index + 1)) {
      context.addIssue({
        code: "custom",
        message: "attempt numbers must be contiguous and ordered",
        path: ["attempts"],
      })
    }
  })

export type TargetedPracticeTrainingRecordTargetRoleWire = z.infer<
  typeof targetedPracticeTrainingRecordTargetRoleSchema
>
export type TargetedPracticeTrainingRecordSetupWire = z.infer<
  typeof targetedPracticeTrainingRecordSetupSchema
>
export type TargetedPracticeTrainingRecordQuestionWire = z.infer<
  typeof targetedPracticeTrainingRecordQuestionSchema
>
export type TargetedPracticeTrainingRecordFollowUpWire = z.infer<
  typeof targetedPracticeTrainingRecordFollowUpSchema
>
export type TargetedPracticeTrainingRecordEvaluationWire = z.infer<
  typeof targetedPracticeTrainingRecordEvaluationSchema
>
export type TargetedPracticeTrainingRecordReviewWire = z.infer<
  typeof targetedPracticeTrainingRecordReviewSchema
>
export type TargetedPracticeTrainingRecordAttemptWire = z.infer<
  typeof targetedPracticeTrainingRecordAttemptSchema
>
export type TargetedPracticeTrainingRecordDetailWire = z.infer<
  typeof targetedPracticeTrainingRecordDetailResponseSchema
>

export const targetedPracticeRecordWireSchema = targetedPracticeTrainingRecordDetailResponseSchema

export const targetedPracticeTrainingRecordSummarySchema = z
  .object({
    recordId: uuidSchema,
    kind: z.literal("targetedPractice"),
    language: interactionLanguageSchema,
    status: z.enum(["completed", "endedEarly", "partiallyCompleted"]),
    startedAt: dateTimeSchema,
    endedAt: dateTimeSchema,
    durationSeconds: z.number().int().nonnegative(),
    targetRole: targetedPracticeTrainingRecordTargetRoleSchema,
    answeredQuestionCount: z.number().int().nonnegative(),
    totalQuestionCount: z.number().int().nonnegative(),
    overallScore: z.number().min(0).max(100).nullable(),
    reviewSummary: z.string().nullable(),
    questionType: practiceQuestionTypeSchema,
    difficulty: practiceDifficultySchema,
  })
  .strict()

export const trainingRecordsPaginationSchema = z
  .object({
    page: z.number().int().min(1),
    pageSize: z.number().int().min(1).max(100),
    totalItems: z.number().int().nonnegative(),
    totalPages: z.number().int().nonnegative(),
  })
  .strict()

export const trainingRecordsPageResponseSchema = z
  .object({
    items: z.array(targetedPracticeTrainingRecordSummarySchema),
    pagination: trainingRecordsPaginationSchema,
  })
  .strict()

export const trainingRecordKindOverviewSchema = z
  .object({
    recordCount: z.number().int().nonnegative(),
    completedRecordCount: z.number().int().nonnegative(),
    averageScore: z.number().min(0).max(100).nullable(),
  })
  .strict()

export const trainingRecordsOverviewResponseSchema = z
  .object({
    totalRecordCount: z.number().int().nonnegative(),
    completedRecordCount: z.number().int().nonnegative(),
    totalDurationSeconds: z.number().int().nonnegative(),
    answeredQuestionCount: z.number().int().nonnegative(),
    averageScore: z.number().min(0).max(100).nullable(),
    targetRoles: z.array(targetedPracticeTrainingRecordTargetRoleSchema),
    byKind: z
      .object({
        targetedPractice: trainingRecordKindOverviewSchema,
        mockInterview: trainingRecordKindOverviewSchema,
      })
      .strict(),
  })
  .strict()

export type TargetedPracticeTrainingRecordSummaryWire = z.infer<
  typeof targetedPracticeTrainingRecordSummarySchema
>
export type TrainingRecordsPaginationWire = z.infer<typeof trainingRecordsPaginationSchema>
export type TrainingRecordsPageWire = z.infer<typeof trainingRecordsPageResponseSchema>
export type TrainingRecordKindOverviewWire = z.infer<typeof trainingRecordKindOverviewSchema>
export type TrainingRecordsOverviewWire = z.infer<typeof trainingRecordsOverviewResponseSchema>

// Record-specific wire names keep the boundary discoverable without duplicating schemas.
export const trainingRecordTargetRoleSchema = targetedPracticeTrainingRecordTargetRoleSchema
export const targetedPracticeSetupWireSchema = targetedPracticeTrainingRecordSetupSchema
export const targetedPracticeQuestionWireSchema = targetedPracticeTrainingRecordQuestionSchema
export const trainingRecordFollowUpWireSchema = targetedPracticeTrainingRecordFollowUpSchema
export const trainingRecordEvaluationWireSchema = targetedPracticeTrainingRecordEvaluationSchema
export const trainingRecordReviewWireSchema = targetedPracticeTrainingRecordReviewSchema
export const targetedPracticeAttemptWireSchema = targetedPracticeTrainingRecordAttemptSchema

File: apps/server/src/riva/schemas/training_records.py
from datetime import datetime
from enum import StrEnum
from typing import Annotated, Literal

from pydantic import ConfigDict, Field, field_validator, model_validator

from riva.core.language import InteractionLanguage
from riva.schemas.base import APIModel
from riva.schemas.evaluation import (
    MAX_PRACTICE_EVALUATION_DIMENSIONS,
    PracticeDimensionScore,
    PracticeEvaluationScore,
)
from riva.schemas.job_description_parsing import Company, RoleTitle
from riva.schemas.practice_recommendation import PracticeRecommendationOutput
from riva.schemas.practice_sessions import (
    MAX_PRACTICE_FOLLOW_UPS,
    PracticeAnswerResponse,
    PracticeFollowUpReferenceAnswerResponse,
    PracticeMainReferenceAnswerResponse,
    PracticeQuestionSource,
)
from riva.schemas.practice_review import (
    MAX_PRACTICE_REVIEW_ITEMS,
    ReviewItem,
    ReviewOverallPerformance,
    ReviewWeakness,
)
from riva.schemas.profile import StandardUUID
from riva.schemas.question_cards import (
    QuestionCardDifficulty,
    QuestionCardPrompt,
    QuestionCardQuestionType,
    QuestionCardTextList,
)


def _validate_aware_timestamp(value: datetime) -> datetime:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError("timestamps must be timezone-aware")
    return value


class TrainingRecordAPIModel(APIModel):
    model_config = ConfigDict(extra="forbid")


class TrainingRecordKind(StrEnum):
    TARGETED_PRACTICE = "targetedPractice"
    MOCK_INTERVIEW = "mockInterview"


class TrainingRecordStatus(StrEnum):
    COMPLETED = "completed"
    ENDED_EARLY = "endedEarly"
    PARTIALLY_COMPLETED = "partiallyCompleted"


class TrainingRecordTargetRoleResponse(TrainingRecordAPIModel):
    id: StandardUUID
    title: RoleTitle
    company: Company


class TargetedPracticeTrainingRecordSummaryResponse(TrainingRecordAPIModel):
    record_id: StandardUUID
    kind: Literal[TrainingRecordKind.TARGETED_PRACTICE]
    language: InteractionLanguage
    status: TrainingRecordStatus
    started_at: datetime
    ended_at: datetime
    duration_seconds: Annotated[int, Field(ge=0)]
    target_role: TrainingRecordTargetRoleResponse
    answered_question_count: Annotated[int, Field(ge=0)]
    total_question_count: Annotated[int, Field(ge=0)]
    overall_score: float | None = Field(default=None, ge=0, le=100)
    review_summary: str | None
    question_type: QuestionCardQuestionType
    difficulty: QuestionCardDifficulty

    _validate_started_at = field_validator("started_at")(_validate_aware_timestamp)
    _validate_ended_at = field_validator("ended_at")(_validate_aware_timestamp)


class TrainingRecordsPaginationResponse(TrainingRecordAPIModel):
    page: Annotated[int, Field(ge=1)]
    page_size: Annotated[int, Field(ge=1, le=100)]
    total_items: Annotated[int, Field(ge=0)]
    total_pages: Annotated[int, Field(ge=0)]


class TrainingRecordsPageResponse(TrainingRecordAPIModel):
    items: list[TargetedPracticeTrainingRecordSummaryResponse]
    pagination: TrainingRecordsPaginationResponse


class TargetedPracticeSetupResponse(TrainingRecordAPIModel):
    source: PracticeQuestionSource
    prioritize_weaknesses: bool


class TargetedPracticeQuestionRecordResponse(TrainingRecordAPIModel):
    question_card_id: StandardUUID
    prompt: QuestionCardPrompt
    question_type: QuestionCardQuestionType
    difficulty: QuestionCardDifficulty
    assessed_capabilities: QuestionCardTextList
    is_saved: bool
    is_marked_weak: bool
    reference_answer: PracticeMainReferenceAnswerResponse


class TrainingRecordFollowUpResponse(TrainingRecordAPIModel):
    question_id: StandardUUID
    prompt: QuestionCardPrompt
    order: Annotated[int, Field(ge=1, le=MAX_PRACTICE_FOLLOW_UPS)]
    asked_at: datetime
    answer: PracticeAnswerResponse | None
    reference_answer: PracticeFollowUpReferenceAnswerResponse

    _validate_asked_at = field_validator("asked_at")(_validate_aware_timestamp)


class TrainingRecordEvaluationResponse(TrainingRecordAPIModel):
    overall_score: PracticeEvaluationScore
    dimension_scores: Annotated[
        list[PracticeDimensionScore],
        Field(
            min_length=4,
            max_length=MAX_PRACTICE_EVALUATION_DIMENSIONS,
        ),
    ]
    evaluated_at: datetime

    _validate_evaluated_at = field_validator("evaluated_at")(
        _validate_aware_timestamp
    )


class TrainingRecordReviewResponse(TrainingRecordAPIModel):
    overall_performance: ReviewOverallPerformance
    highlights: list[ReviewItem] = Field(
        default_factory=list,
        max_length=MAX_PRACTICE_REVIEW_ITEMS,
    )
    main_issues: list[ReviewItem] = Field(
        default_factory=list,
        max_length=MAX_PRACTICE_REVIEW_ITEMS,
    )
    improvement_suggestions: list[ReviewItem] = Field(
        default_factory=list,
        max_length=MAX_PRACTICE_REVIEW_ITEMS,
    )
    reusable_answer_structure: list[ReviewItem] = Field(
        default_factory=list,
        max_length=MAX_PRACTICE_REVIEW_ITEMS,
    )
    exposed_weaknesses: list[ReviewWeakness] = Field(
        default_factory=list,
        max_length=MAX_PRACTICE_REVIEW_ITEMS,
    )


class TargetedPracticeAttemptRecordResponse(TrainingRecordAPIModel):
    attempt_id: StandardUUID
    attempt_number: Annotated[int, Field(ge=1)]
    retry_of_attempt_id: StandardUUID | None
    completed_at: datetime | None
    question: TargetedPracticeQuestionRecordResponse
    main_answer: PracticeAnswerResponse | None
    follow_ups: list[TrainingRecordFollowUpResponse] = Field(default_factory=list)
    evaluation: TrainingRecordEvaluationResponse | None
    review: TrainingRecordReviewResponse | None
    recommendation: PracticeRecommendationOutput | None

    _validate_completed_at = field_validator("completed_at")(
        lambda value: _validate_aware_timestamp(value)
        if value is not None
        else value
    )


class TargetedPracticeTrainingRecordDetailResponse(TrainingRecordAPIModel):
    record_id: StandardUUID
    kind: TrainingRecordKind
    status: TrainingRecordStatus
    language: InteractionLanguage
    started_at: datetime
    ended_at: datetime
    duration_seconds: Annotated[int, Field(ge=0)]
    target_role: TrainingRecordTargetRoleResponse
    setup: TargetedPracticeSetupResponse
    attempts: list[TargetedPracticeAttemptRecordResponse] = Field(
        min_length=1
    )
    exposed_weaknesses: list[ReviewWeakness] = Field(default_factory=list)
    recommendation: PracticeRecommendationOutput | None

    _validate_started_at = field_validator("started_at")(_validate_aware_timestamp)
    _validate_ended_at = field_validator("ended_at")(_validate_aware_timestamp)

    @model_validator(mode="after")
    def validate_timeline_and_attempts(
        self,
    ) -> "TargetedPracticeTrainingRecordDetailResponse":
        if self.ended_at < self.started_at:
            raise ValueError("ended_at cannot be before started_at")
        attempt_numbers = [attempt.attempt_number for attempt in self.attempts]
        if attempt_numbers != list(range(1, len(attempt_numbers) + 1)):
            raise ValueError("attempt numbers must be contiguous and ordered")
        return self


# These aliases make the record-specific names discoverable without creating
# a second wire contract for the same projections.
TrainingRecordQuestionResponse = TargetedPracticeQuestionRecordResponse
TrainingRecordAttemptResponse = TargetedPracticeAttemptRecordResponse
TrainingRecordSummaryResponse = TargetedPracticeTrainingRecordSummaryResponse

__all__ = [
    "TargetedPracticeAttemptRecordResponse",
    "TargetedPracticeQuestionRecordResponse",
    "TargetedPracticeTrainingRecordSummaryResponse",
    "TargetedPracticeSetupResponse",
    "TargetedPracticeTrainingRecordDetailResponse",
    "TrainingRecordAPIModel",
    "TrainingRecordAttemptResponse",
    "TrainingRecordEvaluationResponse",
    "TrainingRecordFollowUpResponse",
    "TrainingRecordKind",
    "TrainingRecordQuestionResponse",
    "TrainingRecordReviewResponse",
    "TrainingRecordSummaryResponse",
    "TrainingRecordStatus",
    "TrainingRecordTargetRoleResponse",
    "TrainingRecordsPageResponse",
    "TrainingRecordsPaginationResponse",
]
